src/Microsoft.IdentityModel.Xml/DsigSerializer.cs

```diff
@@ -196,46 +161,93 @@ public virtual KeyInfo ReadKeyInfo(XmlReader reader)
-        private void ReadCertificate(XmlReader reader, KeyInfo keyInfo)
+        /// <param name="reader">A <see cref="XmlReader"/> positioned on a <see cref="XmlSignatureConstants.Elements.X509Data"/> element.</param>
+        private X509Data ReadX509Data(XmlReader reader)
         {
-            keyInfo.CertificateData = reader.ReadElementContentAsString();
-            var embededCert = new X509Certificate2(Convert.FromBase64String(keyInfo.CertificateData));
-            keyInfo.Kid = embededCert.Thumbprint;
+            var data = new X509Data();
+
+            reader.ReadStartElement(XmlSignatureConstants.Elements.X509Data, XmlSignatureConstants.Namespace);
+            while (reader.IsStartElement())
+            {
+                if (reader.IsStartElement(XmlSignatureConstants.Elements.X509Certificate, XmlSignatureConstants.Namespace))
+                {
+                    data.Certificates.Add(reader.ReadElementContentAsString());
+                }
+
+                else if (reader.IsStartElement(XmlSignatureConstants.Elements.X509IssuerSerial, XmlSignatureConstants.Namespace))
+                {
+                    if (data.IssuerSerial != null)
+                        throw XmlUtil.LogReadException(LogMessages.IDX30015, XmlSignatureConstants.Elements.X509IssuerSerial);
+                    data.IssuerSerial = ReadIssuerSerial(reader);
+                }
+
+                else if (reader.IsStartElement(XmlSignatureConstants.Elements.X509SKI, XmlSignatureConstants.Namespace))
+                {
+                    if (data.SKI != null)
+                        throw XmlUtil.LogReadException(LogMessages.IDX30015, XmlSignatureConstants.Elements.X509SKI);
+                    data.SKI = reader.ReadElementContentAsString();
+                }
+
+                else if (reader.IsStartElement(XmlSignatureConstants.Elements.X509SubjectName, XmlSignatureConstants.Namespace))
+                {
+                    if (data.SubjectName != null)
+                        throw XmlUtil.LogReadException(LogMessages.IDX30015, XmlSignatureConstants.Elements.X509SubjectName);
+                    data.SubjectName = reader.ReadElementContentAsString();
+                }
+
+                else if (reader.IsStartElement(XmlSignatureConstants.Elements.X509CRL, XmlSignatureConstants.Namespace))
+                {
+                    if (data.CRL != null)
+                        throw XmlUtil.LogReadException(LogMessages.IDX30015, XmlSignatureConstants.Elements.X509CRL);
+                    data.CRL = reader.ReadElementContentAsString();
+                }
+
+                else
+                {
+                    // Skip the element since it is not one of  <X509Certificate>, <X509IssuerSerial>, <X509SKI>, <X509SubjectName>, <X509CRL>
+                    Logger.WriteWarning(LogMessages.IDX30300, reader.ReadOuterXml());
+                }
+            }
+
+            // </X509Data>
+            reader.ReadEndElement();
```

Comment:
If we were on an empty element, this will throw.
In that case, we should throw and then we don't need the check below.

Reply:
Done!